solver/src/oneinch_solver.rs

```diff
@@ -4,3 +4,232 @@
 //! single GPv2 order and produce a settlement directly against 1Inch.
 
 pub mod api;
+
+use self::api::{OneInchClient, Slippage, Swap, SwapQuery};
+use crate::{
+    encoding::EncodedInteraction,
+    interactions::Erc20ApproveInteraction,
+    liquidity::{slippage::MAX_SLIPPAGE_BPS, LimitOrder, Liquidity},
+    settlement::{Interaction, Settlement},
+    solver::Solver,
+};
+use anyhow::Result;
+use contracts::{GPv2Settlement, ERC20};
+use ethcontract::{dyns::DynWeb3, U256};
+use futures::future;
+use maplit::hashmap;
+use model::order::OrderKind;
+use rand::seq::SliceRandom as _;
+use std::fmt::{self, Display, Formatter};
+
+/// A GPv2 solver that matches GP **sell** orders to direct 1Inch swaps.
+#[derive(Debug)]
+pub struct OneInchSolver {
+    settlement_contract: GPv2Settlement,
+    client: OneInchClient,
+}
+
+impl OneInchSolver {
+    /// Creates a new 1Inch solver instance for specified settlement contract
+    /// instance.
+    pub fn new(settlement_contract: GPv2Settlement) -> Self {
+        Self {
+            settlement_contract,
+            client: Default::default(),
+        }
+    }
+
+    /// Settles a single sell order against a 1Inch swap.
+    async fn settle_order(&self, order: LimitOrder) -> Result<Settlement> {
+        debug_assert_eq!(
+            order.kind,
+            OrderKind::Sell,
+            "only sell orders should be passed to settle_order"
+        );
+
+        let spender = self.client.get_spender().await?;
+        let sell_token = ERC20::at(&self.web3(), order.sell_token);
+        let existing_allowance = sell_token
+            .allowance(self.settlement_contract.address(), spender.address)
+            .call()
+            .await?;
+
+        let swap = self
+            .client
+            .get_swap(SwapQuery {
+                from_token_address: order.sell_token,
+                to_token_address: order.buy_token,
+                amount: order.sell_amount,
+                from_address: self.settlement_contract.address(),
+                slippage: Slippage::basis_points(MAX_SLIPPAGE_BPS).unwrap(),
+                // Disable balance/allowance checks, as the settlement contract
+                // does not hold balances to traded tokens.
+                disable_estimate: Some(true),
+            })
```

Comment:
I recall that 1inch allows to specify best price or gas efficient as a quote parameter. I would imagine the default is best price and that is also what we would choose on the user's behalf here, but I am not seeing any place where this is passed through to the 1inch client.

Reply:
That flag is currently not supported by the API implementation. We can definitely add it (but in a separate PR I think).

Reply:
https://github.com/gnosis/gp-v2-services/issues/538